### PR TITLE
RFC: secureboot.eclass: easier and more releable signing of bootloaders and kernels

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
This adds a small `secureboot.eclass` that calls `sbsign` to sign efi executables (and can be used for other files as well).

## The problem:

Sure you can manually sign e.g. `systemd-boot`, but when `bootctl update` updates the bootloader files on the ESP the signature is gone and secure boot breaks. If you have enabled the systemd-boot automatic updating service this can come as an unpleasant surprise.

## The solution:

Use the environment variables `SECUREBOOT_SIGN_KEY` and `SECUREBOOT_SIGN_CERT` (similar to `linux-mod-r1.eclass`) to sign any efi executables (or regular kernel files) provided by a package *before* they get installed to the file system. It is now guaranteed that any tooling will install signed files, and as such bootloader and kernel upgrades become more reliable with secure boot enabled.

- `bootctl update` and `bootctl install` are smart enough to install `systemd-bootx64.efi.signed` instead of `systemd-bootx64.efi` if it is present.
- AFAIK `refind-install` doesn't do this so we sign the files in place instead.
- `sys-kernel/gentoo-kernel` (via `dist-kernel-utils.eclass`) is also signed in place because the `90-uki-copy` `kernel-install` plugin enforces the file has the `.efi` extension (not `.efi.signed`). 
- The next release of dracut will expect uki files to be named `uki.efi`, we can already make this change now so we are prepared. 
- Signing uki's was already possible using the dracut configuration flags `uefi_secureboot_cert` and `uefi_secureboot_key`. Now we can also sign regular kernel files, this is required to use secure boot in combination with grub since uki support for grub hasn't landed in a released version yet.
- Tried to make the same changes to the grub ebuild, but apparently there is no grub efi executable generated during compilation. Presumably this is done by `grub-install`? So there are still some extra steps required here to make secure boot work with grub.

## Still todo:

- Somehow make `systemd-boot` work with `shim`. With these changes we can enable secure boot provided we ensure the key we use is registered with the uefi firmware. However,  It is not always possible or desirable to enrol custom keys. Shim makes it possible to use secure boot without enrolling custom keys but is currently hardcoded to load grub. `refind-install` is smart enough to deal with this if you call it with the `--shim` option but getting this to work with `systemd-boot` requires manual copying which is undesirable because then the bootloader doesn't get updated automatically any more. Possibly the best solution here is an upstream solution (https://github.com/systemd/systemd/issues/27234).
- Somehow also provide a way to semi-automatically sign the grub efi executable.

Lets discuss this with the relevant package/eclass maintainers first before I send it to the mailing list. @floppym @gentoo/base-system @sveyret @gentoo/proxy-maint @gentoo/systemd @gentoo/dist-kernel @mgorny

### CODE REVIEW

**Comment by floppym** [comment]
The systemd changes seem fine to me.

Given that "secureboot" is an opt-in USE flag, I think it would be reasonable to have pkg_setup die instead of warn if the user hasn't configured the variables properly.

File: profiles/arch/amd64/use.mask
Comment (ionenwks) [inline profiles/arch/amd64/use.mask:9]: I'd change the messages on arches where it's being unmasked, it sounds confusing when it says Mask but then it's being unmasked.

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:119]: Should use an array for `"${args[@]}"` (i.e. SECUREBOOT_SIGN_KEY could potentially have spaces in the path and it'd work fine with an array), and more quoting to prevent splitting on input/output file too.

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:105]: quotes here too

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:117]: Wouldn't hurt to be consistent for `|| die \` vs `|| <newline> die`, you've used the latter in other spots which is fine.

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:157]: Feel free to do how you want in your own eclass (just style), but fwiw *all* quotes here are unnecessary except the one on the die "".

Not needed for the `path="${ED}"` either up either. Quotes don't matter for bash variable assignment even if the variable contain spaces, but they do if you were doing `args=( not-really-a-var="something" )`

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:168]: `EXPORT_FUNCTIONS` should be after the closing `fi` at the very bottom. This matters with `inherit` ordering, i.e. if inherited last then it should use its pkg_setup. But if another eclass inherited it first then the meaning unexpectedly changes given EXPORT won't be ran again.

File: eclass/secureboot.eclass
Comment (ionenwks) [inline eclass/secureboot.eclass:170]: Most eclasses set this right at the start if the `if`  block rather than bottom, wouldn't hurt to be consistent with these.

**Review by ionenwks — COMMENTED** [review]
Only had a quick glance and haven't tried, so just simple surface stuff I've seen.

File: eclass/dist-kernel-utils.eclass
Comment (ionenwks) [inline eclass/dist-kernel-utils.eclass:44]: Tend to prefer `type -P` over `command -v` when it's bash (latter for posix), the former won't pickup functions should they exist and behaves more like calling `which` would.

File: eclass/dist-kernel-utils.eclass
Comment (ionenwks) [inline eclass/dist-kernel-utils.eclass:46]: Double space before sed, the quotes (`="$()"`) are also unneeded.

What's that `2>&1` for? You need to pickup sed error messages inside the layout variable?

Comment (AndrewAmmerlaan) [reply eclass/dist-kernel-utils.eclass:46]: This is a remnant from some testing earlier that I forgot to remove

File: eclass/secureboot.eclass
Comment (AndrewAmmerlaan) [reply eclass/secureboot.eclass]: Yeah I have a habit of using too many quotes, once I get in the flow I forget that bash is not python and that not every single string needs quotes.

File: eclass/secureboot.eclass
Comment (AndrewAmmerlaan) [reply eclass/secureboot.eclass]: Oops, good catch, thanks!

File: eclass/dist-kernel-utils.eclass
Comment (AndrewAmmerlaan) [reply eclass/dist-kernel-utils.eclass]: Fixed! Thanks for the review

**Comment by AndrewAmmerlaan** [comment]
Checked the portage file list and added similar changes for some other packages that install efi files.

`sys-boot/shim` is a special case, the shim efi executable is already signed with the Microsoft keys (that is the whole point of shim after all) but we still need to sign the MokManager that comes with it since it is by default only signed with the Fedora keys (where we borrow these files from). So to make it work we would either need to enrol the Fedora keys in the Mok database, or simply sign the files with the user specified key using `secureboot.eclass`. 

Also CC'ing the maintainers of these other packages @ZeroChaos- @chithanh @tamiko @gentoo/virtualization 

Maybe there is no point in signing the `sys-firmware/edk2-ovmf` efi files since they are used for virtual machines only. On the other hand, it can't hurt to sign them and maybe it is nice to have the option to add (additional) signatures. WDYT?

**Comment by thesamesam** [comment]
@zippy2 as well might be interested.

**Comment by ZeroChaos-** [comment]
I maintain shim and understanding how shim works is the limit of my understanding of secure boot stuff.  Please feel free to do whatever you think improves usability with secure boot enabled.  I am happy to discuss specific ideas if you want to "run it by someone", but if you have an obvious improvement consider this an official ACK from me.

**Comment by AndrewAmmerlaan** [comment]
Just minor changes in this push:
- `find` now finds `.efi` files case-insensitive because sometimes efi executables have the `.EFI` extension.
- Fixed a QA warning about a missing `eend` if a file was already signed.

**Review by xen0n — COMMENTED** [review]
I've skimmed through all changes. They look reasonable to me but I'd leave the green tick to folks more familiar with Secure Boot (I always disable it the moment any box arrives at my home 😹).

**Comment by AndrewAmmerlaan** [comment]
Some more small changes in this push:
- `REQUIRED_USE="secureboot? ( modules-sign )"` for kernels. For the kernel to successfully boot with secure boot enabled there must be support for signed kernel modules and all necessary kernel modules must have a valid signature.
- Moved the code that checks if the key and certificate exist into `_secureboot_die_if_unset` so we already die in `pkg_setup` if the variables are set to non-existent keys.
- Dropped the commit that enforces a match between the dracut uefi setting and the kernel-install layout setting. After thinking about this some more I concluded this approach is wrong because the next release of systemd will include a kernel-install plugin for ukify. If the user wishes to use this plugin then dracut has to be configured to only build an initrd (i.e. `uefi=no`). This is a valid workflow and enforcing this match as I was doing here would break this.

**Comment by AndrewAmmerlaan** [comment]
Eclass patches have been sent to the mailing list for further review. I've omitted the patches to the individual packages to not spam everyone with too much email.

File: eclass/secureboot.eclass
Comment (mgorny) [inline eclass/secureboot.eclass:154]: Is there a valid case for calling this function when `${ED}` does not exist?

File: eclass/kernel-build.eclass
Comment (mgorny) [inline eclass/kernel-build.eclass:32]: I don't exactly like having two control vars that pretty much will always be used together but I guess at this point you'd have to rename the old one for this to make sense with just one var.

File: eclass/secureboot.eclass
Comment (mgorny) [inline eclass/secureboot.eclass:138]: FWICS you're not using the `--ext` variant anywhere in this PR. Is there any use case for it right now? If not, let's keep things simple and drop it until it's actually necessary.

File: eclass/secureboot.eclass
Comment (ionenwks) [reply eclass/secureboot.eclass:154]: fwiw it allows to give the error just below rather than have `find` itself ungracefully fail on missing path

Comment (ionenwks) [reply eclass/secureboot.eclass:154]: Albeit I guess this is partially copied from linux-mod-r1, except in linux-mod-r1 you could specify the path rather than it always be `${ED}`

File: eclass/secureboot.eclass
Comment (AndrewAmmerlaan) [reply eclass/secureboot.eclass:113]: Right, I guess this does not really make sense indeed.

File: eclass/secureboot.eclass
Comment (AndrewAmmerlaan) [reply eclass/secureboot.eclass:154]: Yeah this was copied from `linux-mod-r1.eclass. But this check isn't really necessary in our case indeed.

File: eclass/kernel-build.eclass
Comment (AndrewAmmerlaan) [reply eclass/kernel-build.eclass:32]: Yeah me neither, I did it like this anyway because it feels a bit weird to use a var here that we have defined and documented in another eclass.

I'd be open to doing this differently though, but I don't really see a proper alternative.

File: eclass/secureboot.eclass
Comment (AndrewAmmerlaan) [reply eclass/secureboot.eclass:138]: Not really. I'll just drop it then.

**Comment by AndrewAmmerlaan** [comment]
Found some more packages installing efi executables.

The efi files from `sys-apps/memtest86+` have a `.efi32` and `.efi64` suffix, so I adjusted the `find` in the eclass to also account for these cases. No further changes.

**Comment by AndrewAmmerlaan** [comment]
v2 has been sent to the mailinglist, let me know if there are any further comments on the eclass patches.

I'll update the relevant wiki pages after the eclass patches are merged.

File: eclass/kernel-build.eclass
Comment (mgorny) [inline eclass/kernel-build.eclass:414]: Sort by execution order, please.

**Comment by AndrewAmmerlaan** [comment]
Eclass patches have been merged, and PR has been rebased. Some patches to individual packages still remaining.

@chithanh @tamiko @hydrapolic @zmedico @robbat2 @benkohler @ConiKost Let me know if these are okay to merge too.


Now to update the wiki.

**Review by ajakk — APPROVED** [review]
Sorry for the late review, but this looks really good, and thank you for working on it!

**Pull request closed without merging**
